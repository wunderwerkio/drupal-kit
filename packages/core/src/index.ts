export * from "./Drupalkit";
export * from "./types";
