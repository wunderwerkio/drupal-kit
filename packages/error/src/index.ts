import * as DrupalkitTypes from "@drupalkit/types";

import { RequestErrorOptions as DrupalkitErrorOptions } from "./types";

/**
 * Custom error class to help with error handling.
 */
export class DrupalkitError extends Error {
  /**
   * Error name.
   */
  name: "HttpError";

  /**
   * HTTP status code.
   */
  status: number;

  /**
   * Request options that lead to the error.
   */
  request: DrupalkitTypes.RequestOptions;

  /**
   * Response object if a response was received.
   */
  response?: DrupalkitTypes.DrupalkitResponse<unknown>;

  /**
   * Construct a new DrupalkitError.
   *
   * @param message - Error message.
   * @param statusCode - HTTP status code.
   * @param options - Request / Response that lead to the error.
   */
  constructor(
    message: string,
    statusCode: number,
    options: DrupalkitErrorOptions,
  ) {
    super(message);

    // Maintains proper stack trace (only available on V8)
    // @ts-ignore
    if (Error.captureStackTrace) {
      // @ts-ignore
      Error.captureStackTrace(this, this.constructor);
    }

    this.name = "HttpError";
    this.status = statusCode;

    if (options.response) {
      this.response = options.response;
    }

    // redact request credentials without mutating original request options
    const requestCopy = Object.assign({}, options.request);
    if (options.request.headers.authorization) {
      requestCopy.headers = Object.assign({}, options.request.headers, {
        authorization: options.request.headers.authorization.replace(
          / .*$/,
          " [REDACTED]",
        ),
      });
    }

    this.request = requestCopy;
  }
}
