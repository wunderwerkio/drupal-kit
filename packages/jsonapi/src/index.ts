export * from "./resources";
export * from "./DrupalkitJsonApi";
export * from "./DrupalkitJsonApiError";
