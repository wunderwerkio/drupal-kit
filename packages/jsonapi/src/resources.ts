import { DrupalJsonApiParamsInterface } from "drupal-jsonapi-params";
import { LinkObject, ResourceObject, Response } from "ts-json-api";

/* eslint-disable jsdoc/require-description-complete-sentence */
/* eslint-disable @typescript-eslint/no-empty-interface */

/**
 * Interface describing all valid JSON:API resources for the .resource()
 * method.
 *
 * This interface is meant to be augmented via declaration merging
 * in your own code!
 *
 * Below is an example of how the interface is expected to be used.
 * The key is a string literal of the entity type and bundle type separated
 * by double dashes.
 *
 * The value is a object with the following properties:
 *   - `resource`: Type of the JSON:API resource object.
 *   - `operations`: Union of string literals of allowed operations.
 *
 * The resource type MUST extend the `ResourceObject` interface from the
 * `ts-json-api` package.
 *
 * Valid operations:
 *   - readSingle
 *   - readMany
 *   - create
 *   - update
 *   - delete
 *
 * @example
 * ```ts
 * declare module "@drupal-kit/jsonapi" {
 *   interface JsonApiResources {
 *     "file--file": {
 *       resource: FileResource;
 *       operations: "readSingle" | "readMany" | "create" | "update" | "delete";
 *     }
 *   }
 * }
 * ```
 */
export interface JsonApiResources {}

export interface JsonApiIndex extends Response<[]> {
  links: {
    [type: string]: LinkObject;
  };
}

/**
 * Utility types.
 */

/**
 * Extract the create payload type from a resource object.
 */
type ResourceCreatePayload<R extends ResourceObject> = {
  data: Pick<R, "type" | "attributes" | "relationships"> &
    Partial<Pick<R, "id">>;
};

/**
 * Extract the update payload type from a resource object.
 */
type ResourceUpdatePayload<R extends ResourceObject> = {
  data: Pick<R, "id" | "type"> & {
    attributes?: Partial<R["attributes"]>;
    relationships?: Partial<R["relationships"]>;
  };
};

/**
 * Parameters for various operations.
 */

type ReadParameters = {
  queryParams?: DrupalJsonApiParamsInterface;
};

export type ReadSingleParameters = ReadParameters & {
  uuid: string;
};

export type ReadManyParameters = ReadParameters;

export type CreateParameters<Resource extends ResourceObject> = {
  payload: ResourceCreatePayload<Resource>;
};

export type UpdateParameters<Resource extends ResourceObject> = {
  uuid: string;
  payload: ResourceUpdatePayload<Resource>;
};

export type DeleteParameters = {
  uuid: string;
};

/**
 * Infer correct parameters by operation and resource type.
 */
export type ToParameters<
  Operation,
  Resource extends ResourceObject,
> = "readSingle" extends Operation
  ? ReadSingleParameters
  : "readMany" extends Operation
  ? ReadManyParameters
  : "create" extends Operation
  ? CreateParameters<Resource>
  : "update" extends Operation
  ? UpdateParameters<Resource>
  : "delete" extends Operation
  ? DeleteParameters
  : never;

export type ToReturnType<
  Operation,
  Resource extends ResourceObject,
> = "readSingle" extends Operation
  ? ReadSingleParameters
  : "readMany" extends Operation
  ? ReadManyParameters
  : "create" extends Operation
  ? CreateParameters<Resource>
  : "update" extends Operation
  ? UpdateParameters<Resource>
  : "delete" extends Operation
  ? DeleteParameters
  : never;
