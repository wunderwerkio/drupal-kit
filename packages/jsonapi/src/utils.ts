import { RequestOptions } from "@drupalkit/types";
import { getHeader } from "@drupalkit/utils";

/**
 * Checks if request is a JSON:API request.
 *
 * @param req - The request options.
 */
export const isJsonApiRequest = (req: RequestOptions) => {
  const contentType = getHeader("content-type", req);

  return contentType === "application/vnd.api+json";
};
