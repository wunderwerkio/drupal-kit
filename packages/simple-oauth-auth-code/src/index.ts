export * from "./DrupalkitSimpleOauthAuthCode.js";
export * from "./types.js";
