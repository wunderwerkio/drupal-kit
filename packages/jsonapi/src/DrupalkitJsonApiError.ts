import { DrupalkitError, DrupalkitErrorOptions } from "@drupal-kit/error";

type JsonApiError = {
  title: string;
  status: string;
  detail: string;
  source?: {
    pointer?: string;
  };
};

/**
 * Enhanced DrupalkitError with JSON:API specific data.
 */
export class DrupalkitJsonApiError extends DrupalkitError {
  /**
   * Array of JSON:API errors.
   */
  readonly errors: JsonApiError[];

  /**
   * @inheritdoc
   */
  constructor(
    message: string,
    statusCode: number,
    errors: JsonApiError[],
    options: DrupalkitErrorOptions,
  ) {
    super(message, statusCode, options);

    this.errors = errors;
  }

  /**
   * Checks if this error instance contains validation errors.
   */
  public hasValidationErrors() {
    return this.errors.some((error) => error.status == "422");
  }

  /**
   * Get an array of all fields with invalid values.
   */
  public getInvalidFields() {
    return this.getErrorsByStatus(422)
      .map((error) => {
        const pointer = error?.source?.pointer;

        if (!pointer) {
          return null;
        }

        return pointer.split("/").pop();
      })
      .filter((field) => field !== null);
  }

  /**
   * Get all errors with the given status code.
   *
   * @param statusCode - The status code to filter by.
   */
  public getErrorsByStatus(statusCode: number) {
    return this.errors.filter(
      (error) => error.status.toString() == statusCode.toString(),
    );
  }

  /**
   * Create new error instance from a DrupalkitError.
   *
   * @param error - The DrupalkitError to create an instance from.
   */
  public static fromDrupalkitError(error: DrupalkitError) {
    const errors = error.response
      ? this.extractJsonApiErrorsFromResponse(error.response.data)
      : [];

    return new DrupalkitJsonApiError(error.message, error.statusCode, errors, {
      request: error.request,
      response: error.response,
    });
  }

  /**
   * Extract JSON API errors from response data.
   *
   * @param data - The response data to extract errors from.
   */
  public static extractJsonApiErrorsFromResponse(
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    data: any,
  ) {
    if (!data?.errors) {
      return [];
    }

    return data.errors as JsonApiError[];
  }
}
