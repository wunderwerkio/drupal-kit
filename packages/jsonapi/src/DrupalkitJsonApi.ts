import { ResourceObject, Response } from "ts-json-api";
import { Err, Ok } from "ts-results";
import { Drupalkit, DrupalkitOptions, Query } from "@drupalkit/core";

import { DrupalkitJsonApiError } from "./DrupalkitJsonApiError";
import {
  JsonApiIndex,
  JsonApiResources,
  ReadManyParameters,
  ReadSingleParameters,
  ToParameters,
} from "./resources.js";
import { isJsonApiRequest } from "./utils";

declare module "@drupalkit/core" {
  interface DrupalkitOptions {
    jsonApiPrefix?: string;
  }
}

/**
 * Drupalkit plugin that provides integration with the Drupal JSON:API.
 *
 * @param drupalkit - The Drupalkit instance.
 * @param drupalkitOptions - The options for the Drupalkit instance.
 */
export const DrupalkitJsonApi = (
  drupalkit: Drupalkit,
  drupalkitOptions: DrupalkitOptions,
) => {
  const defaultHeaders: HeadersInit = {
    Accept: "application/vnd.api+json",
    "Content-Type": "application/vnd.api+json",
  };

  /**
   * Retrieves the JSON:API index.
   *
   * @returns A result object containing the JSON:API index or an error.
   */
  const getIndex = async () => {
    const url = buildJsonApiUrl("");

    const response = await drupalkit.request<JsonApiIndex>(url, {
      method: "GET",
      headers: {
        ...defaultHeaders,
      },
    });

    if (response.err) {
      return response;
    }

    return Ok(response.val.data);
  };

  /**
   * Retrieves a single JSON:API resource object.
   *
   * @param type - The type of resource object to retrieve.
   * @param parameters - The parameters to use for the query.
   * @param options - Optional settings to override locale and default locale.
   * @param options.localeOverride - An optional override for the locale.
   * @param options.defaultLocaleOverride - An optional override for the default locale.
   * @returns A result object containing the resource object or an error.
   */
  const getResource = async <R extends ResourceObject>(
    type: R["type"],
    parameters: ReadSingleParameters,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
    },
  ) => {
    const path = type.replace("--", "/") + "/" + parameters.uuid;

    const url = buildJsonApiUrl(path, {
      ...options,
      query: parameters.queryParams?.getQueryObject(),
    });

    const result = await drupalkit.request<Response<R>>(url, {
      method: "GET",
      headers: defaultHeaders,
    });

    if (result.err) {
      return result;
    }

    return Ok(result.val.data);
  };

  /**
   * Retrieves multiple JSON:API resource objects.
   *
   * @param type - The type of resource object to retrieve.
   * @param parameters - The parameters to use for the query.
   * @param options - Optional settings to override locale and default locale.
   * @returns A result object containing the resource object or an error.
   */
  const getResourceCollection = async <R extends ResourceObject>(
    type: R["type"],
    parameters: ReadManyParameters,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
    },
  ) => {
    const path = type.replace("--", "/");

    const url = buildJsonApiUrl(path, {
      ...options,
      query: parameters.queryParams?.getQueryObject(),
    });

    const result = await drupalkit.request<Response<R[]>>(url, {
      method: "GET",
      headers: defaultHeaders,
    });

    if (result.err) {
      return result;
    }

    return Ok(result.val.data);
  };

  /**
   * Constructs a JSON API URL for use with Drupal.
   *
   * @param path - The path to the JSON API endpoint.
   * @param options - An optional object containing additional options.
   * @param options.localeOverride - An optional override for the locale.
   * @param options.defaultLocaleOverride - An optional override for the default locale.
   * @returns The constructed URL as a string.
   */
  const buildJsonApiUrl = (
    path: string,
    options?: {
      localeOverride?: string;
      defaultLocaleOverride?: string;
      query?: Query;
    },
  ) => {
    const prefix = drupalkitOptions.jsonApiPrefix ?? "jsonapi";

    return drupalkit.buildUrl(path, {
      ...options,
      customPrefix: prefix,
    });
  };

  /**
   * Create DrupalkitJsonApiError for JSON:API failed requests.
   */
  drupalkit.hook.error("request", (error) => {
    // Only care about JSON:API requests.
    if (isJsonApiRequest(error.request)) {
      throw DrupalkitJsonApiError.fromDrupalkitError(error);
    }

    throw error;
  });

  /**
   * Extend the Drupalkit instance.
   */
  return {
    jsonApi: {
      buildJsonApiUrl,
      getIndex,
      async resource<
        Type extends keyof JsonApiResources,
        Resource extends JsonApiResources[Type]["resource"],
        Operation extends JsonApiResources[Type]["operations"],
        Params extends ToParameters<Operation, Resource>,
        Return extends Record<
          Operation,
          "readSingle" extends Operation
          ? Awaited<ReturnType<typeof getResource>>
          : "readMany" extends Operation
          ? Awaited<ReturnType<typeof getResourceCollection>>
          : Err<Error>
        >,
      >(
        type: Type,
        operation: Operation,
        parameters: Params,
        options?: {
          localeOverride?: string;
          defaultLocaleOverride?: string;
        },
      ): Promise<Return[Operation]> {
        switch (operation) {
          case "readSingle":
            return (await getResource(
              type,
              parameters as ReadSingleParameters,
              options,
            )) as Return[Operation];

          case "readMany":
            return (await getResourceCollection(
              type,
              parameters as ReadManyParameters,
              options,
            )) as Return[Operation];
        }

        return Err(
          new Error(`Unknown operation "${operation}"`),
        ) as Return[Operation];
      },
    },
  };
};
