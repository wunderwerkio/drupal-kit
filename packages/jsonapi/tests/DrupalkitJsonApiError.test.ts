import { DrupalkitError } from "@drupalkit/error";

import { DrupalkitJsonApiError } from "../src";
import JsonApiErrorResponse from "./fixtures/jsonapi_multiple_errors.json";

describe("DrupalkitJsonApiError", () => {
  const request = {
    method: "GET",
    headers: {
      "Content-Type": "application/vnd.api+json",
    },
  };

  it("should instanciate from DrupalkitError", () => {
    const error = new DrupalkitError("test-error", 400, {
      request,
    });

    const jaError = DrupalkitJsonApiError.fromDrupalkitError(error);

    expect(jaError).toBeInstanceOf(DrupalkitJsonApiError);
    expect(jaError).toBeInstanceOf(DrupalkitError);
  });

  it("should extract errors from JSON:API response", () => {
    const error = new DrupalkitError("test-error", 400, {
      request,
      response: {
        headers: {},
        status: 422,
        data: JsonApiErrorResponse,
        url: "some-url"
      }
    });

    const jaError = DrupalkitJsonApiError.fromDrupalkitError(error);

    expect(jaError.hasValidationErrors()).toBeTruthy()
    expect(jaError.getInvalidFields()).toEqual(["field_firstname", "field_lastname"]);
    expect(jaError.getErrorsByStatus(400)).toHaveLength(1);
    expect(jaError.getErrorsByStatus(422)).toHaveLength(2);

    expect(jaError.getErrorsByStatus(400)).toMatchSnapshot("bad-request-errors");
    expect(jaError.getErrorsByStatus(422)).toMatchSnapshot("unprocessable-entity-errors");
  });
});
