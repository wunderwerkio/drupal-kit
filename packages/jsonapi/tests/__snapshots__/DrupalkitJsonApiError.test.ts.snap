// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrupalkitJsonApiError should extract errors from JSON:API response: bad-request-errors 1`] = `
[
  {
    "detail": "\`wrong-field\` is not a valid relationship field name. Possible values: node_type, revision_uid, uid, menu_link, field_media.",
    "links": {
      "info": {
        "href": "http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1",
      },
      "via": {
        "href": "https://my-drupal.com/jsonapi/node/article/5f5f5f5f-5f5f-5f5f-5f5f-5f5f5f5f5f5f",
      },
    },
    "status": "400",
    "title": "Bad Request",
  },
]
`;

exports[`DrupalkitJsonApiError should extract errors from JSON:API response: unprocessable-entity-errors 1`] = `
[
  {
    "detail": "field_firstname: Dieser Wert sollte nicht leer sein.",
    "source": {
      "pointer": "/data/attributes/field_firstname",
    },
    "status": "422",
    "title": "Unprocessable Entity",
  },
  {
    "detail": "field_lastname: Dieser Wert sollte nicht leer sein.",
    "source": {
      "pointer": "/data/attributes/field_lastname",
    },
    "status": "422",
    "title": "Unprocessable Entity",
  },
]
`;
